keps/sig-windows/2258-node-service-log-viewer/README.md

```diff
@@ -0,0 +1,375 @@
+- [x] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+A Kubernetes cluster administrator has to log in to the relavant control-plane
+or worker nodes to view the logs of the API server, kubelet etc. Or they would
+have to implement a client side reader. A simpler and more elegant method would
+be to allow them to use the kubectl CLI to also view these logs similar to
+using it for other interactions with the cluster. Given the sensitive nature of
+the information in node logs, this feature will only be available to cluster
+administrators.
+
+## Motivation
+
+Troubleshooting issues with control-plane and worker nodes typically requires 
+a cluster administrator to SSH into the nodes for debugging. While certain
+issues will require being on the node, issues with the kube-proxy or kubelet,
+to name a couple, could be solved by perusing their logs. However this
+too requires the administrator to SSH access into the nodes. Having a way for
+them to view the logs using kubectl will significantly simplify their
+troubleshooting.
+
+
+### Goals
+Provide a cluster administrator with a streaming view of logs using kubectl
+without them having to implement a client side reader or logging into the node.
+This would work for:
+- Services on Linux worker and control plane nodes:
+  - That have systemd / journald support.
+  - That have services that log to `/var/log/`
+- Windows worker nodes (all supported variants) that log to `C:\var\log`,
+  System and Application logs, Windows Event Logs and Event Tracing (ETW).
+
+### Non-Goals
+- Providing support for non-systemd Linux distributions.
+- Reporting logs for nodes that have config or connection issues with the
+  cluster.
+- Getting logs from services that do not use /var/log/.
+
+## Proposal
+
+### Implement client for logs endpoint viewer (OS agnostic)
+- Extend `kubectl logs` to work with node objects.
+- Implement a client for the `/var/log/` kubelet endpoint viewer.
+
+### Linux distros with systemd / journald
+Supplement the the `/var/log/` endpoint viewer on the kubelet with a thin shim
+over the `journal` directory that shells out to journalctl. Then extend
+`kubectl logs` to also work with node objects.
+
+### Linux distributions without systemd / journald
+Running the new "kubectl logs nodes" command against services on nodes that do
+not use systemd / journald should return "OS not supported". However getting
+logs from `/var/log/` should work on all systems.
```

Comment:
This is an optional endpoint today and kubelets can disable it by setting `enableSystemLogHandler: false` in kubelet config. Would this feature be disabled by doing that as well?

Reply:
That's a good point. Yes, I think it makes sense that this feature would get disabled by doing that. I will call that out in the KEP. Do we need the `NodeLogs` feature gate in addition to `enableSystemLogHandler`?

Reply:
There are three varieties of existing clusters:
1. Existing clusters that set `enableSystemLogHandler: false`. They clearly don't want this feature.
2. Existing clusters that are ok with exposing /var/log but wouldn't want this new feature (maybe there's more sensitive stuff in journald or windows system logs)
3. Existing clusters that are ok with exposing /var/log and with this new feature.

Describing what each type of existing cluster should do to get its desired result would be helpful.

Feature gates are not a long-term enablement/disablement mechanism. Is the intent to allow surfacing /var/log but not this new feature long-term, or are you expecting anyone exposing /var/log today will be required to enable this new feature when it reaches GA?

Reply:
@SaranBalaji90 @sjenning  it looks like you authored and/or approved https://github.com/kubernetes/kubernetes/pull/87273 which added this field.

Can either of you comment here?
Should we also add a new flag to enable / disable exposing journald / windows system logs via `kubectl logs nodes` or should we just re-use the same flag and disable this functionality if `enabledSystemLoghandler=false`?

Reply:
The meaning of `enabledSystemLoghandler` seems to encompass both types of logs (`/var/log` and `journal`).  

So my thought at the moment is that in the long term (the `NodeLogs` feature is GA) a cluster that has `enableSystemLogHandler: true` will enable both the `/var/log/` and the `journal` endpoints. In the short term a cluster that wants to enable this feature needs to set `enableSystemLogHandler: true` and enable the `NodeLogs` feature flag. 

WDYT?

Reply:
I was thinking the same thing.

Reply:
Thanks for the clear explanation @liggitt. I have added our current approach to the KEP.